### ISSUE TITLE
[bug] WebSQL adapter not detected correctly in pouchdb-adapter-websql/src/valid.js (when SQLite Plugin 2 + Cordova used)

### ISSUE THREAD

**Issue by dharders** [pouchdb/pouchdb#5417]
### Environment 

Hybrid App

- cordova 6.2.0
- cordova-plugin-wkwebview-engine
- cordova-labs-local-webserver
- cordova-plugin-sqlite-2

### Browser/platform 

iOS 9.3.2 WKWebView via cordova plugin

### Adapter 

WebSQL - via cordova-plugin-sqlite-2

### Server CouchDB

CouchDB

### Please describe your issue.

I have a hybrid app running successfully on iOS 9.3.2 using the faster WKWebView and PouchDB 5.3.2 ( `{adapter: 'websql'}` and  **cordova-plugin-sqlite-2**. All good.

If I change nothing else but bump my `package.json` pouchdb version to `5.4.x` and then `npm install` and build/run my app on my device I get the error: **Error: Adapter is missing**

I had to dig for the error using Safari Dev tools and found it in the `taskqueue` of my local db object. i.e. 

```javascript
var db = new PouchDB('mydb', {adapter: 'websql', revs_limit: 100, auto_compaction: true});
console.log(db);    // db.taskqueue.failed : 'Error: Adapter is missing'
```

If I swap back to 5.3.2, all is good and app works perfectly.

I noticed there were some [deprecations in 5.4.0](https://github.com/pouchdb/pouchdb/releases/tag/5.4.0) but not sure if these are to blame. I'm more suspicious of bugfix #5197 / PR #5198. I get why this fix was introduced for browser validation, but with cordova + sqlite plugin 2 we can get it to work. Are we artificially disabling websql here for cordova apps too ?

**Edit:** Turns out we are: [valid.js#L56](https://github.com/pouchdb/pouchdb/blob/master/packages/pouchdb-adapter-websql/src/valid.js#L56) assumes a `SQLitePlugin` object is defined before `deviceready`.

But with **cordova-plugin-sqlite-2** the object isn't available until the `deviceready` event fires and it is different casing `sqlitePlugin` i.e. there is no `SQLitePlugin` object ever.

I fixed my problem by adding `return true;` at the beginning of `function valid()` so it is definitely an artificial PouchDB imposed limitation for Cordova users. Now I can run PouchDB 5.4.4 (with the above mod to valid.js) on iOS 9.3.2 using the faster WKWebView  + cordova-plugin-sqlite-2. Boom.

This may fix @tropperstyle 's issue in #5386 too.

**Comment by nolanlawson** [comment]
This is a nasty bug. I don't understand how it was even working before, since `valid()` is, in theory, run at exactly the same time. Perhaps there was some subtle timing issue that changed.

Unfortunately I don't see an easy way to resolve this without loading PouchDB _after_ the `'deviceready'` event, or hacking `valid()` as you did.

Question: Are you calling `new PouchDB()` _after_ `'deviceready'` or before? Because ideally the `valid()` function should be re-run when you call that.

**Comment by nolanlawson** [comment]
Also if someone could provide a test case to reproduce, that'd be great. "Hello world" Cordova app would do.

**Comment by nolanlawson** [comment]
Ah, I see that https://github.com/pouchdb/pouchdb/issues/5197 was definitely the cause of this bug. We no longer detect WKWebView as valid because it has no WebSQL.

**Comment by nolanlawson** [comment]
I think ultimately what this comes down to is that `valid()` has to be run at the moment PouchDB is istantiated. Either that, or we need a new adapter plugin _just_ for the SQLite Plugin, or we need to distribute PouchDB as its own Cordova plugin. AFAIK there is no way to know whether or not a Cordova plugin "will" be loaded before `'deviceready'` has happened.

**Comment by dharders** [comment]
@nolanlawson Note that even after `deviceready` the global object `SQLitePlugin` that valid() is looking  for is still not defined. `sqlitePlugin` is though

**Comment by dharders** [comment]
I guess a workaround for (Cordova) users could be to define a global object `var SQLitePlugin = {}` BEFORE the pouchdb.js script import.

It's a bit of a hack, but if documented could be OK if we must detect at the time PouchDB is instantiated.

**Comment by dharders** [comment]
@nolanlawson 

> Are you calling new PouchDB() after 'deviceready' or before? 

I've tried both. Same result because...

> even after deviceready the global object SQLitePlugin that valid() is looking for is still not defined. sqlitePlugin is though

**Comment by dharders** [comment]
> Also if someone could provide a test case to reproduce, that'd be great. "Hello world" Cordova app would do.

I'll knock one up now... BRB

**Comment by dharders** [comment]
@nolanlawson 

Here is the [test case](https://github.com/dharders/pouchdb-cordova-wkwebview-issue)

**Comment by dharders** [comment]
OK I've spent a full day trying to detect the `websql` adapter **before** `deviceready` and I can't find a way. I've come to the conclusion that PouchDB would have to first detect if cordova is present by `if (typeof cordova !== 'undefined')...` then re-run adapter processing on `deviceready`. 

This gets messy as currently there is caching of previous results to work around too and how would we guarantee that the PouchDB hook is run before the users `deviceready` call to `new PouchDB(...)` etc

**Best solution without touching PouchDB source:** 

Add this to [Common Errors](https://pouchdb.com/errors.html) / docs

```html
<!-- Cordova iOS WKWebView + sqlitePlugin 2 + PouchDB Fix -->

<head>
<script type="text/javascript" src="cordova.js"></script>
<script>
  if (typeof cordova !== 'undefined' && cordova.platformId === 'ios') {
    var SQLitePlugin = {} ;    // work-around for iOS WKWebView + sqlitePlugin 2 + PouchDB 
  }
</script>
<script type="text/javascript" src="pouchdb.js"></script> <!-- must be after SQLitePlugin defined -->
</head>

<!-- Ensure you call var db = new PouchDB('db', {adapter: 'websql'}); only after 'deviceready' event -->
```

**Comment by nolanlawson** [comment]
Oh lol, I didn’t even notice the `SQLitePlugin` vs `sqlitePlugin` difference. Dumb typo on my part.

I’d be in favor of just adding `typeof cordova !== 'undefined'` to the `valid()` function. If a user is on Cordova, they can always install the SQLite Plugin.

**Comment by nolanlawson** [comment]
^ proposed fix

**Comment by dharders** [comment]
Fix looks good to me. 

It's a tricky one to write a test for though... maybe mock openDatabase() and throw DOM Exception 18, and similar with IDB ? Define a global Cordova object ? Dunno ...

**Comment by daleharvey** [comment]
Without actually testing against the platforms the mocks will just confirm that the code is doing what we wrote, without getting cordova in CI manually verifying is about the best we can do, hopefully `idb-next` sorts all this out :) fixed in https://github.com/pouchdb/pouchdb/commit/f05ec7cc265a67c27fd89baa5d5e00c8a1e9a95c

**Comment by nevace** [comment]
Thanks so much for this, it was driving me mad!

**Comment by oliviertassinari** [comment]
I'm facing the same issue using the `5.4.5` release.
I'm not sure the issue is completely solved.

The following code snippet is unpredictable. It works **20%** of the time:
```js
import PouchDB from 'pouchdb';

document.addEventListener('deviceready', () => {
  dbLocal = new PouchDB('db', {
    adapter: 'websql',
  });
  dbLocal.info().then(console.log.bind(console));
}, false);
```
I'm having the following error: `Adapter is missing`, where `PouchDB.adapters[opts.adapter]` only contains `https` and `http`

On the other hand, this code snipped works **100%** of the time:
```js
document.addEventListener('deviceready', () => {
  const PouchDB = require('pouchdb');

  dbLocal = new PouchDB('db', {
    adapter: 'websql',
  });
  dbLocal.info().then(console.log.bind(console));
}, false);
```

**Comment by nevace** [comment]
Is it not an issue with the way Babel is transpiling it? What does the transpiled code look like? I'm assuming you're using Angular 1 with Babel.

I manually call my function and have no issue...

```this.setDatabase = function(databaseName) {
        this.db = new PouchDB(databaseName, { adapter: 'websql' });
      }.bind(this);```

**Comment by oliviertassinari** [comment]
I don't think that it has anything to do with babel as it's transpiled with a require.
What seems to matter most is when the require is done and when the sql cordova plugin is available
in the `winow`.
Wait a second. I'm lady loading the `cordova.js` script with an `async` html property.
`window.cordova` may be undefined when I'm requiring pouchdb in my main script.

**Comment by dharders** [comment]
Sounds like your unpredictable results are due to the async loading. Update your index.html to contain <script src="cordova.js"></script> in your header before any other script imports (there is no real penalty as the file is served locally). If your still having issues, post a more complete sample (ie your above sample looks like you're calling dblocal.info() before the deviceready event has a chance to fire)

**Comment by oliviertassinari** [comment]
>  (there is no real penalty as the file is served locally)

I'm not sure about this statement. What about the parsing and the evaluation phase of the js? The network is just one step. 
I'm gonna investigate a bit more on the issue.

**Comment by dharders** [comment]
I should clarify to say no real 'noticeable' penalty. ( in my testing anyway). I was more interested in whether the modification fixed your issue.

It's a requirement to have cordova instantiated **before** pouchdb to work with wkwebview/sqlite 2 so I don't think you're gonna get around it, unless you can somehow guarantee the order with lazy loading (which I'm unaware of due to the async nature)

If this fixed your issue, please start a new 'performance' issue (probably in Cordova repo) so we don't get off topic here. But if your issue still remains, please keep the conversation going so we can fix it for everyone.

**Comment by oliviertassinari** [comment]
@dharders I have more update on this issue. I confirm the flaky success are linked to a `window.cordova` not being defined depending on the order of the loading of my assets.
My workaround seems the best solution available.

Regarding the performances of using `async`. I had forgotten to put the main script `async` too. That was blocking the rendering. After derering this script and the CSS, the first paint is around 200ms instead of 1.5s. That's much better.

**Comment by nolanlawson** [comment]
Hm, that's a problem. In my fix I assumed that `window.cordova` was defined from the very beginning. If that's not the case for Cordova apps, then I don't even see how we can possibly solve this issue. We have to detect that we're running in Cordova in order to fix it.

Another option is to just `window.cordova = true` in the `<head>`, but that's a sad workaround.

Isn't it a best practice to include `cordova.js` before any other dependencies? Did you have the correct ordering and yet the `window.cordova` variable was still flakey?

**Comment by nolanlawson** [comment]
Also this issue has been closed for a long time; if there is still a bug, please reopen a new bug with a reproducible test case so we can track it there; it makes it easier to follow what's going on. :)

**Comment by guyromb** [comment]
Issue still exists with 5.4.5

**Comment by guyromb** [comment]
I fixed it by making sure cordova plugins are loaded before angular:

`
angular
        .element(document)
        .ready(_onDocumentReady)
    ;
function _onDocumentReady() {
        // Wait on cordova (if it is available)
        if (!!(window.cordova)) {
            console.log('Cordova detected, waiting on deviceready');
            document.addEventListener('deviceready', _bootstrapDocument, false);
        }
        else {
            // No cordova, continuing
            _bootstrapDocument();
        }
    }

    function _bootstrapDocument() {
        console.log('------ BOOTSTRAPPING DOCUMENT ------');
        angular.bootstrap(document, ['chime']);
    }
`

**Comment by nolanlawson** [comment]
Folks, this problem has been solved by [pouchdb-adapter-cordova-sqlite](http://github.com/nolanlawson/pouchdb-adapter-cordova-sqlite). It's far too difficult to try to maintain WebSQL validation code in the same place as the Cordova SQLite code, so I split it up into a separate module.

Please use that separate module instead. SQLite support will be dropped from PouchDB core in v6.0.0 entirely.

**Comment by oliviertassinari** [comment]
@nolanlawson It's great to see that PouchDB has embraced the modular approach 👍 .

I have migrated my codebase to use [pouchdb-adapter-cordova-sqlite](https://github.com/nolanlawson/pouchdb-adapter-cordova-sqlite).
That hasn't changed much the approach. We still have to wait for the native sqlite (`cordova.sqlitePlugin`) to be available, hence using:
```js
document.addEventListener('deviceready', () => {
```
That's already in the docs. We are good here.

**Comment by Arxi** [comment]
Sorry to revive this thread, but I can't seem to make sense of this. I use PouchDB and Ionic 1.3.2 and until PouchDB version 5.4.4, this worked perfectly for me with `adapter: 'websql'`
- while using `ionic serve` - i.e. in browser - the plugin would correctly choose WebSQL which made the rapid development in browser possible
- on real device, while using `ionic run android`, the plugin would detect SQLite

After upgrading to 5.4.5 I get the error that I need one of the three sqlite plugins, even though I use Cordova sqlite storage and do the PouchDB init stuff in ionic.Platform.ready().

My question is: can I achieve the same adapter auto-detect functionality by using `pouchdb-adapter-cordova-sqlite` or do I have to stick to PouchDB 5.4.4?

**Comment by nolanlawson** [comment]
@Arxi you can check if `window.cordova` is truthy or not and then just switch between different adapters

```
var db = window.cordova ?
  new PouchDB('mydb.db', {adapter: 'cordova-sqlite'}) :
  new PouchDB('mydb'); // uses idb, falls back to websql
```

**Comment by simran100** [comment]
Hi,I am getting PouchDb adapter undefined Error while running my SAPUI5 project on windows 8.1 phone via visual studio. Can someone please help me out,what adapter should i use to run my project on windows 8.1 phone
